src/vcpkg/base/system.mac.cpp

```diff
@@ -0,0 +1,251 @@
+            {
+                if (ptr) freeifaddrs(ptr);
+            }
+        } interfaces;
+
+        if (getifaddrs(&interfaces.ptr) < 0)
+        {
+            return "0";
+        }
+
+        unsigned char bytes[MAC_BYTES_LENGTH];
+        for (auto interface = interfaces.ptr; interface; interface = interface->ifa_next)
+        {
+            // The ifa_addr field points to a structure containing the interface
+            // address.  (The sa_family subfield should be consulted to
+            // determine the format of the address structure.)  This field may
+            // contain a null pointer.
+            if (!interface->ifa_addr || interface->ifa_addr->sa_family != AF_TYPE ||
+                (interface->ifa_flags & IFF_LOOPBACK))
+            {
+                continue;
+            }
+
+            // Convert the generic sockaddr into a specified representation
+            // based on the value of sa_family, on macOS the AF_PACKET
+            // family is not available so we fall back to AF_LINK.
+            // AF_PACKET and sockaddr_ll: https://man7.org/linux/man-pages/man7/packet.7.html
+            // AF_LINK and sockaddr_dl: https://illumos.org/man/3SOCKET/sockaddr_dl
+#if defined(__linux__)
+            auto address = reinterpret_cast<sockaddr_ll*>(interface->ifa_addr);
+            if (address->sll_halen != MAC_BYTES_LENGTH) continue;
+            std::memcpy(bytes, address->sll_addr, MAC_BYTES_LENGTH);
+#elif defined(__APPLE__)
+            auto address = reinterpret_cast<sockaddr_dl*>(interface->ifa_addr);
+            if (address->sdl_alen != MAC_BYTES_LENGTH) continue;
+            // The macro LLADDR() returns the start of the link-layer network address.
+            std::memcpy(bytes, LLADDR(address), MAC_BYTES_LENGTH);
+#endif
+            auto maybe_mac = mac_bytes_to_string(Span<unsigned char>(bytes, MAC_BYTES_LENGTH));
+            if (is_valid_mac_for_telemetry(maybe_mac))
+            {
+                return Hash::get_string_hash(maybe_mac, Hash::Algorithm::Sha256);
+            }
+        }
+#else
+        // fallback when getifaddrs() is not available
+        struct socket_guard
+        {
+            int fd = -1;
+            ~socket_guard()
+            {
+                if (fd > 0) close(fd);
+            }
+            inline operator int() { return fd; }
+        } fd{socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)};
+
+        if (fd == -1) return "0";
+
+        ifconf interfaces;
+        std::memset(&interfaces, 0, sizeof(ifconf));
+        // Calling ioctil with ifconf.ifc_req set to null returns
+        // the size of buffer needed to contain all interfaces
+        // in ifconf.ifc_len.
+        // https://www.man7.org/linux/man-pages/man7/netdevice.7.html
+        if (ioctl(fd, SIOCGIFCONF, &interfaces) < 0) return "0";
+
+        auto data = std::vector<char>(interfaces.ifc_len);
+        interfaces.ifc_len = data.size();
+        interfaces.ifc_buf = data.data();
```

Comment:
```suggestion
        // add one to ensure that even if remainder > 0, we reserve enough space
        std::vector<ifreq> data(interfaces.ifc_len / sizeof(ifreq) + 1);
        interfaces.ifc_req = data.data();
```
Writing the pointer to `.ifc_buf` and then reading from `.ifc_req`, while _probably_ fine, gives me aliasing shivers. We're assuming this is a vector of `ifreq` anyway.

Reply:
I thought `ifreq` was one of those variable-length structure thingies?

Reply:
I guess not...